### PR TITLE
Handle read date in the Goodreads importer

### LINKED ISSUE #7399: Cannot import reading dates exported from Goodrads
I've exported my reading log from Goodreads using their "Export Library" functionality. However, my "Date Read" column is not being translated into my OL reading log as a "Check-In Date". It would be nice to incorporate that feature to automatically load my reading log history rather than manually updating it for each book.

### PR DESCRIPTION
**Pull request by BoboTiG**
Closes #7399

- Fix for missing read dates when importing Goodreads data.
- Fix the "Import N Books" title that wasn't updated when (un)checking checkboxes.
- Fix the progress bar not taking into account unchecked books in percent computation.

### Technical

Nothing.

### Testing

Open the browser at the `/account/import` page, select your Goodreads export, then click on import.
From there, play with checkboxes to see that the number of books-to-import is updated in real-time.
finally, run the import, and check any read book page to see the check-in date properly set.

### Screenshot

The video of the title being updated:

https://user-images.githubusercontent.com/2033598/216003419-0e79b782-144b-4675-901b-956b52b4af99.mp4

The screenshot with calls to update the read date (it includes several books to cover all use case like unchecked book, and book not in the library):
Screenshot from 2023-02-01 10-18-07

### CODE REVIEW

**Comment by BoboTiG** [comment]
Ready for review.

File: openlibrary/plugins/openlibrary/js/goodreads_import.js [hunk 1]
```diff
@@ -89,39 +101,61 @@ export function initGoodreadsImport() {
...
                                     },
-                                    datatype: 'json',
+                                    dataType: 'json',
                                     success: function () {
-                                        $(`[isbn=${value['ISBN']}]`).append('<td class="success-imported">Imported</td>')
-                                        $(`[isbn=${value['ISBN']}]`).removeClass('selected');
+                                        if (value['Date Read'] !== '') {
+                                            let date_read = value['Date Read'].split('/'); // Format: "YYYY/MM/DD"
+                                            $.ajax({
+                                                url: `${obj['works'][0].key}/check-ins`,
```
Comment (BoboTiG) [inline openlibrary/plugins/openlibrary/js/goodreads_import.js:109]: Is it preferable to use
```
`${obj['works'][0].key}/bookshelvers_events.json`
```
?

Comment (BoboTiG) [reply openlibrary/plugins/openlibrary/js/goodreads_import.js:109]: I'll change, and review the code again. There might be a way to clean-up a little bit that code.

**Comment by mekarpeles** [comment]
@BoboTiG thank you for submitting this improvement! It looks like it's in the right direction. Tagging @jimchamp for a more thorough review.

**Comment by BoboTiG** [comment]
Gentle ping :)

File: openlibrary/plugins/openlibrary/js/goodreads_import.js [hunk 2]
```diff
@@ -2,41 +2,37 @@ import Promise from 'promise-polyfill';
...
-        l = $('.add-book[checked*="checked"]').length;
+        const l = $('.add-book[checked*="checked"]').length;
         $('.import-submit').attr('value', `Import ${l} Books`);
     });
 
     function func1(value) {
-        l = $('.add-book[checked*="checked"]').length;
-        elem = document.getElementById('myBar');
+        var l = $('.add-book[checked*="checked"]').length;
+        var elem = document.getElementById('myBar');
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/goodreads_import.js:35]:
```suggestion
        const l = $('.add-book[checked*="checked"]').length;
        const elem = document.getElementById('myBar');
```
These are not reassigned, so we can declare as a constant.

File: openlibrary/plugins/openlibrary/js/goodreads_import.js [hunk 3]
```diff
@@ -52,76 +48,108 @@ export function initGoodreadsImport() {
...
         prevPromise = Promise.resolve();
         $('input.add-book').each(function () {
             var input = $(this),
                 checked = input.prop('checked');
             var value = JSON.parse(input.val().replace(/'/g, '"'));
             var shelf = value['Exclusive Shelf'];
-            const shelves = {read: 3, 'currently-reading': 2,  'to-read': 1};
             var shelf_id = 0;
-            if (shelves[shelf]){
+            var has_failure = function () {
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/goodreads_import.js:60]:
```suggestion
            const hasFailure = function () {
```
Camel case should be used for Javascript variable names.

File: openlibrary/plugins/openlibrary/js/goodreads_import.js [hunk 4]
```diff
@@ -52,76 +48,108 @@ export function initGoodreadsImport() {
...
+                return $(`[isbn=${value['ISBN']}]`).hasClass('import-failure');
+            };
+            var fail = function (reason) {
+                if (!has_failure()) {
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/goodreads_import.js:64]:
```suggestion
            const fail = function (reason) {
                if (!hasFailure()) {
```

File: openlibrary/plugins/openlibrary/js/goodreads_import.js [hunk 5]
```diff
@@ -52,76 +48,108 @@ export function initGoodreadsImport() {
...
-                                    },
-                                    fail: function () {
-                                        $(`[isbn=${value['ISBN']}]`).append('<td class="error-imported">Error</td><td class="error-imported">Failed to add Rating</td>')
-                                        $(`[isbn=${value['ISBN']}]`).removeClass('selected');
-                                    }
-                                });
+            if (shelf_id === 0) {
+                fail('Book in different Shelf');
+                func1(++count);
+                return true;
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/goodreads_import.js:82]:
```suggestion
                return false;
```
JQuery documentation recommends returning false to exit `.each()` early.

File: openlibrary/plugins/openlibrary/js/goodreads_import.js [hunk 6]
```diff
@@ -52,76 +48,108 @@ export function initGoodreadsImport() {
...
+                if (!has_failure()) {
+                    const element = $(`[isbn=${value['ISBN']}]`);
+                    element.append(`<td class="error-imported">Error</td><td class="error-imported">${reason}</td>'`)
+                    element.removeClass('selected');
+                    element.addClass('import-failure');
+                }
+            };
+
+            if (!checked) {
+                return true;
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/goodreads_import.js:73]:
```suggestion
                return false;
```

File: openlibrary/plugins/openlibrary/js/goodreads_import.js [hunk 7]
```diff
@@ -52,76 +48,108 @@ export function initGoodreadsImport() {
...
-                    $(`[isbn=${value['ISBN']}]`).append('<td class="error-imported">Error</td><td class="error-imported">Book not in collection</td>')
-                    $(`[isbn=${value['ISBN']}]`).removeClass('selected');
-                    func1(++count);
+                        });
+                    }
                 });
-            }
-            else if (checked && shelf_id === 0) {
-                $(`[isbn=${value['ISBN']}]`).append('<td class="error-imported">Error</td><td class="error-imported">Book in different Shelf</td>');
+                if (!has_failure()) {
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/goodreads_import.js:142]:
```suggestion
                if (!hasFailure()) {
```

**Review by jimchamp — APPROVED** [review]
@BoboTiG apologies for the massive delay.  Thanks for refactoring this code!  It should make this easier to maintain in the future.

Have tested that this works as expected.

**Comment by BoboTiG** [comment]
Ah, thanks @jimchamp for taking care of those changes, and the review 🍾

**Pull request merged**